### DISCUSSION TITLE
Updated Dockerfile

### DISCUSSION THREAD

**Discussion by andrewschmidt-a** [shadow/shadow#1015] [Q&A] [answered]
**State your question concisely**
Could someone look at my Dockerfile and tell me if I installed shadow correctly?
https://github.com/andrewschmidt-a/shadow-docker

**Describe your current understanding of Shadow as it relates to your question**
Currently there is some docker layers that are published as a tarball. Its running a really old version of fedora container and I needed to install some newer packages to work alongside of shadow. When I realized I was practically updating the whole os in my docker image I decided to start with a new base image. 

I couldn't find the dockerfile that georgetown used to create the original base layer so I just recreated it from the steps.

**Describe which parts you are unsure about**
Not sure if I got all the sysctl settings configured right. I couldnt actually set them in the docker image itself (permission denied)

Also I have the proposed hotfix for https://github.com/shadow/shadow/issues/805 included

**Comment by stevenengler** [comment, marked as answer]
You're right that the Docker instructions are very out of date, and need to be updated.

> I couldn't find the dockerfile that georgetown used to create the original base layer so I just recreated it from the steps.

I don't think that image was created from a Dockerfile.

> Could someone look at my Dockerfile and tell me if I installed shadow correctly?

It looks okay to me. The main question is, do the tests pass and does Shadow run? If so, then it's probably good. The sysctl settings affect the kernel and usually need to be made outside of the container, so you can't put them in a Dockerfile. Some sysctl options are namespaced so you might be able to set some per-container, but I don't know which ones.

I'm not sure if you're asking for general suggestions to improve your Dockerfile, but if you are, here are some suggestions:

1. Your Dockerfile skips running Shadow's tests. The tests are very important, especially when running in a container, but maybe these are better run on each user's system after they start a container.
2. You may want to `rm -r shadow/build` after you install it to reduce the number of files in your image.
3. It's usually best to reduce the number of RUN steps to reduce the number of layers.
4. When you download an external file in your Dockerfile (for example with wget), you should check the hash to make sure it's the expected file and not a wrong or malicious file. This also helps make your Dockerfile reproducible.
5. Many of your RUN steps run under the root user when they don't need to. For example you build tgen as the root user, when it might be better to build it as your shadow user.
6. I don't think the `&& source /home/shadow/.bashrc` in your Dockerfile does anything, so could be removed.
6. You install the same package in multiple locations, and you install some that I don't think are needed (such as the `kernel-devel` package).

**Reply by ToJen** [reply]
Any idea how to create a new `Dockerfile` based off the current _main_ branch (which I assume has the latest stable changes)?

**Reply by stevenengler** [reply]
The latest stable version is 1.14.0, and the 'main' branch contains some additional changes that haven't made it to a stable release yet. We also have a 'dev' branch which contains much larger changes, and will hopefully be merged into the 'main' branch within the next few weeks.

If you're running large experiments we recommend installing Shadow directly on the host, but if you want to use Docker to just try Shadow or run some smaller tests, here's an example Dockerfile that should help:

```dockerfile
FROM ubuntu:18.04

RUN DEBIAN_FRONTEND=noninteractive apt-get update \
  && DEBIAN_FRONTEND=noninteractive apt-get install -y \
    cmake \
    gcc \
    g++ \
    libc-dbg \
    libglib2.0-dev \
    libigraph-dev \
    make \
    python \
    python-pyelftools \
    xz-utils \
    python-numpy \
    python-lxml \
    python-matplotlib \
    python-networkx \
    python-scipy \
    dstat \
    git \
    htop \
    screen

RUN git clone https://github.com/shadow/shadow.git \
  && cd shadow \
  && git checkout v1.14.0

RUN cd shadow \
  && ./setup build --clean --test \
  && ./setup install

ENV PATH="/root/.shadow/bin:${PATH}"
```

(If you checkout the main branch instead, replace the "python" dependencies with "python3".

To build the image run:

```bash
sudo docker build -t shadow-classic .
```

To start the container run:

```bash
sudo docker run --tmpfs /dev/shm:rw,nosuid,nodev,exec,size=1g --rm -it shadow-classic bash
```

To run Shadow's tests, run the following inside the container:

```bash
cd /shadow
./setup test
```
